src/backend/distributed/planner/recursive_planning.c

```diff
@@ -252,10 +314,515 @@ RecursivelyPlanSubqueriesAndCTEs(Query *query, RecursivePlanningContext *context
+ * that are not on the distribution key.
+ * *
+ * Note that at the point that this function is called, we've already recursively planned all
+ * the leaf subqueries. Thus, we're actually checking whether the joins among the subqueries
+ * on the distribution key or not.
+ */
+static bool
+ShouldRecursivelyPlanNonColocatedSubqueries(Query *subquery,
+											RecursivePlanningContext *context)
+{
+	/*
+	 * If the input query already contains the equality, simply return since it is not
+	 * possible to find any non colocated subqueries.
+	 */
+	if (context->queryContainsDistributionKeyEquality)
+	{
+		return false;
+	}
+
+	/*
+	 * This check helps us in two ways:
+	 *   (i) We're not targeting queries that don't include subqueries at all,
+	 *       they should go through regular planning.
+	 *  (ii) Lower level subqueries are already recursively planned, so we should
+	 *       only bother non-colocated subquery joins, which only happens when
+	 *       there are subqueries.
+	 */
+	if (SubqueryEntryList(subquery) == NIL && SublinkList(subquery) == NIL)
+	{
+		return false;
+	}
+
+	/* direct joins with local tables are not supported by any of Citus planners */
+	if (FindNodeCheckInRangeTableList(subquery->rtable, IsLocalTableRTE))
+	{
+		return false;
+	}
+
+	/*
+	 * Finally, check whether this subquery contains distribution key equality or not.
+	 */
+	if (!SubqueryContainsDistributionKeyEquality(subquery,
+												 context->plannerRestrictionContext))
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * RecursivelyPlanNonColocatedSubqueries gets a query which includes one or more
+ * other subqueries that are not joined on their distribution keys. The function
+ * tries to recursively plan some of the subqueries to make the input query
+ * executable by Citus.
+ *
+ * The function picks an anchor subquery and iterates on the remaining subqueries.
+ * Whenever it finds a non colocated subquery with the anchor subquery, the function
+ * decides to recursively plan the non colocated subquery.
+ *
+ * The function first handles subqueries in FROM clause (i.e., jointree->fromlist) and then
+ * subqueries in WHERE clause (i.e., jointree->quals).
+ *
+ * The function does not treat outer joins seperately. Thus, we might end up with
+ * a query where the function decides to recursively plan an outer side of an outer
+ * join (i.e., LEFT side of LEFT JOIN). For simplicity, we chose to do so and handle
+ * outer joins with a seperate pass on the join tree.
+ */
+static void
+RecursivelyPlanNonColocatedSubqueries(Query *subquery, RecursivePlanningContext *context)
+{
+	NonColocatedSubqueryContext nonColocatedPlanningContext;
+
+	RangeTblEntry *anchorRangeTblEntry = NULL;
+
+	FromExpr *joinTree = subquery->jointree;
+	PlannerRestrictionContext *restrictionContext = NULL;
+
+	/* we couldn't pick an anchor subquery, so we cannot continue */
+	anchorRangeTblEntry = AnchorRte(subquery);
+	if (anchorRangeTblEntry == NULL)
```

Comment:
Is this something we could move into `CreateNonColocatedSubqueryJoinContext`, given that we already pass a `subquery`? Mostly asking because it makes it easier to move `CreateNonColocatedSubqueryJoinContext` into a different file.